crates/router/src/lib.rs

```diff
@@ -65,4 +66,86 @@ impl Guest for Component {
     }
 }
 
+/// Apply any request transformations needed for the given route.
+fn apply_request_transformations(
+    request: IncomingRequest,
+    raw_route: String,
+) -> anyhow::Result<IncomingRequest> {
+    let headers_to_add = calculate_default_headers(&request, raw_route)?
+        .into_iter()
+        .flat_map(|(k, v)| {
+            k.iter()
+                .map(move |s| (s.to_string(), v.clone().into_bytes()))
+        })
+        .chain(request.headers().entries());
+    let headers = Headers::new();
+    for (key, value) in headers_to_add {
+        headers.append(&key, &value).unwrap();
+    }
+    let new = OutgoingRequest::new(headers);
+    let _ = new.set_scheme(request.scheme().as_ref());
+    let _ = new.set_authority(request.authority().as_deref());
+    let _ = new.set_method(&request.method());
+    let _ = new.set_path_with_query(request.path_with_query().as_deref());
+    Ok(new_request(new, Some(request.consume().unwrap())))
+}
+
+const FULL_URL: &[&str] = &["SPIN-FULL-URL", "X-FULL-URL"];
+const PATH_INFO: &[&str] = &["SPIN-PATH-INFO", "PATH-INFO"];
+const MATCHED_ROUTE: &[&str] = &["SPIN-MATCHED-ROUTE", "X-MATCHED-ROUTE"];
+const COMPONENT_ROUTE: &[&str] = &["SPIN-COMPONENT-ROUTE", "X-COMPONENT-ROUTE"];
+const RAW_COMPONENT_ROUTE: &[&str] = &["SPIN-RAW-COMPONENT-ROUTE", "X-RAW-COMPONENT-ROUTE"];
+const BASE_PATH: &[&str] = &["SPIN-BASE-PATH", "X-BASE-PATH"];
+const CLIENT_ADDR: &[&str] = &["SPIN-CLIENT-ADDR", "X-CLIENT-ADDR"];
+/// Calculate the default headers for the given request.
+fn calculate_default_headers<'a>(
+    req: &IncomingRequest,
+    raw_route: String,
+) -> anyhow::Result<Vec<(&'a [&'a str], String)>> {
+    let mut res = vec![];
+    // TODO: calculate base path from manifest
+    let base = "/".to_owned();
+    let abs_path = req.path_with_query().expect("cannot get path and query");
+    let scheme = req.scheme();
+    let scheme = match scheme
+        .as_ref()
+        .unwrap_or(&bindings::wasi::http::types::Scheme::Https)
+    {
+        bindings::wasi::http::types::Scheme::Http => "http",
+        bindings::wasi::http::types::Scheme::Https => "https",
+        bindings::wasi::http::types::Scheme::Other(s) => s,
+    };
+    let host = req
+        .headers()
+        .get(&"host".to_owned())
+        .into_iter()
+        .find(|v| !v.is_empty())
+        .map(String::from_utf8)
+        .transpose()?
+        .unwrap_or_else(|| "localhost".to_owned());
+
+    let matched_route =
+        spin_http::routes::RoutePattern::sanitize_with_base(base.clone(), raw_route.clone());
+
+    let path_info = spin_http::routes::RoutePattern::from(base.clone(), raw_route.clone())
+        .relative(&abs_path)?;
+    res.push((PATH_INFO, path_info));
+    let full_url = format!("{}://{}{}", scheme, host, abs_path);
+    res.push((FULL_URL, full_url));
+    res.push((MATCHED_ROUTE, matched_route));
+
+    res.push((BASE_PATH, "TODO".to_owned()));
+    res.push((RAW_COMPONENT_ROUTE, "TODO".to_owned()));
```

Comment:
Is `RAW_COMPONENT_ROUTE` here not just `raw`?

Reply:
Whoops, yes it is.